crates/via-router/src/visitor.rs

```diff
@@ -103,41 +107,52 @@ impl<'a, 'b, T> Visitor<'a, 'b, T> {
         node: &'a Node<T>,
     ) {
         let (start, end) = path_segment_range;
+        let is_exact_match = self.path.get(depth + 1).is_none();
         let path_segment_value = &self.path.value[start..end];
 
         for index in node.entries() {
             let next = &self.routes[*index];
 
-            if path_segment_value != next.pattern {
-                // The path segment does not match the pattern of the next node.
-                // We can skip this node and continue to search for a match.
-                continue;
-            }
-
-            if let Pattern::CatchAll(_) = next.pattern {
-                // The end offset of `path_segment` should be the end offset
-                // of the last path segment in the url path.
-                let path_segment_range = self.path.slice_from(depth);
-
-                // The next node has a `CatchAll` pattern and will be considered
-                // an exact match. Due to the nature of `CatchAll` patterns, we
-                // do not have to continue searching for descendants of this
-                // node that match the remaining path segments.
-                self.matches.push(Match {
-                    path_segment_range,
-                    is_exact_match: true,
-                    node: next,
-                });
-            } else {
-                self.matches.push(Match {
-                    path_segment_range,
-                    is_exact_match: self.path.get(depth + 1).is_none(),
-                    node: next,
-                });
-
-                // Continue to match descendants of `next` against the path
-                // segment at the next depth.
-                self.match_at_depth(depth + 1, next);
+            match next.pattern {
```

Comment:
Switching to `match` resulted in the most significant performance improvement. On my machine, the `find_all_matches` benchmark takes around 140 nanoseconds per iteration, down from approximately 170 nanoseconds.